metals/src/main/scala/scala/meta/internal/metals/ConnectionProvider.scala

```diff
@@ -296,34 +299,103 @@ class ConnectionProvider(
   }
 
   object Connect {
+    class RequestInfo(val request: ConnectRequest) {
+      val promise: Promise[BuildChange] = Promise()
+      val cancelPromise: Promise[Unit] = Promise()
+      def cancel(): Boolean = cancelPromise.trySuccess(())
+    }
+
+    @volatile private var currentRequest: Option[RequestInfo] = None
+    private val queue = new ConcurrentLinkedQueue[RequestInfo]()
+
+    def getOngoingRequest(): Option[RequestInfo] = currentRequest
+
     def connect[T](request: ConnectRequest): Future[BuildChange] = {
-      request match {
-        case Disconnect(shutdownBuildServer) => disconnect(shutdownBuildServer)
-        case Index(check) => index(check)
-        case ImportBuildAndIndex(session) => importBuildAndIndex(session)
-        case ConnectToSession(session) => connectToSession(session)
-        case CreateSession(shutdownBuildServer) =>
-          createSession(shutdownBuildServer)
-        case GenerateBspConfigAndConnect(buildTool, shutdownServer) =>
-          generateBspConfigAndConnect(buildTool, shutdownServer)
-        case BloopInstallAndConnect(
-              buildTool,
-              checksum,
-              forceImport,
-              shutdownServer,
-            ) =>
-          bloopInstallAndConnect(
-            buildTool,
-            checksum,
-            forceImport,
-            shutdownServer,
-          )
+      val info = addToQueue(request)
+      pollAndConnect()
+      info.promise.future
+    }
+
+    private def addToQueue(request: ConnectRequest): RequestInfo =
+      synchronized {
+        val info = new RequestInfo(request)
+        val iter = queue.iterator()
+        while (iter.hasNext()) {
+          val curr = iter.next()
+          request.cancelCompare(iter.next().request) match {
```

Comment:
Could you explain the logic a bit better here? Could we use some case objects instead of just numbers?